toolkit/offchain/src/register.rs

```diff
@@ -0,0 +1,362 @@
+#![allow(dead_code)]
+
+use crate::csl::{
+	get_first_validator_budget, InputsBuilderExt, OgmiosUtxoExt, TransactionBuilderExt,
+	TransactionContext,
+};
+use crate::plutus_script::PlutusScript;
+use anyhow::anyhow;
+use cardano_serialization_lib::{
+	BigNum, DataCost, ExUnits, JsError, MinOutputAdaCalculator, Transaction, TransactionBuilder,
+	TransactionOutputBuilder, TxInputsBuilder,
+};
+use ogmios_client::{
+	query_ledger_state::QueryLedgerState, query_network::QueryNetwork, transactions::Transactions,
+	types::OgmiosUtxo,
+};
+use partner_chains_plutus_data::registered_candidates::block_producer_registration_to_plutus_data;
+use sidechain_domain::{BlockProducerRegistration, McTxHash, UtxoId};
+
+pub async fn register<C: QueryLedgerState + QueryNetwork + Transactions>(
+	genesis_utxo: UtxoId,
+	block_producer_registration: &BlockProducerRegistration,
+	input_utxo: UtxoId,
+	payment_signing_key: [u8; 32],
+	ogmios_client: &C,
+) -> anyhow::Result<McTxHash> {
+	let ctx = TransactionContext::for_payment_key(payment_signing_key, ogmios_client).await?;
+	let validator = crate::scripts_data::registered_candidates_scripts(genesis_utxo)?;
+	let validator_address = validator.address_bech32(ctx.network)?;
+	let input_utxo = ctx
+		.payment_key_utxos
+		.iter()
+		.find(|u| u.to_domain() == input_utxo)
+		.ok_or(anyhow!("input utxo not found at payment address"))?;
+	let own_registration_utxos = ogmios_client.query_utxos(&[validator_address]).await?;
```

Comment:
I don't fully understand this line here. This queries for all registrations under the validator address, not only the transaction submitter's own ones, right? What will happen with utxos of other candidates?

Reply:
Good point! I checked and it turns out that there was another filtering step in the original code here that I missed. Will implement now.

Reply:
I've missed that as well in review, even though I have my deregister draft that uses this filtering by `own_pkh`

Reply:
I'm moving this logic to the next PR.